pkg/controller/common/dnsentrieshelper.go

```diff
@@ -107,3 +107,22 @@ func (h *ShootDNSEntriesHelper) DeleteAll() error {
 	}
 	return h.client.DeleteAllOf(h.ctx, &dnsapi.DNSEntry{}, client.InNamespace(h.ext.Namespace), matchingLabel)
 }
+
+// ForceDeleteAll forces deletion of DNSEntries by removing the finalizers first.
+// Warning: calling this method can result in leaked DNS record sets in the infrastructure and should only be used as last resort.
+func (h *ShootDNSEntriesHelper) ForceDeleteAll() error {
+	entries, err := h.List()
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range entries {
+		patch := client.MergeFrom(entry.DeepCopy())
+		entry.SetFinalizers(nil)
+		if err := client.IgnoreNotFound(h.client.Patch(h.ctx, &entry, patch)); err != nil {
+			return fmt.Errorf("removing finalizers for DNSEntry %s failed: %w", entry.Name, err)
+		}
+	}
+
+	return h.DeleteAll()
```

Comment:
We should delete the resource first and then remove the finalizer, other wise some controller can add the finalizers in between the step of removing finalizer and then deleting the resource.